node/subsystem/dispatch-gen/src/lib.rs

```diff
@@ -0,0 +1,216 @@
+struct EnumVariantDispatchWithTy {
+	// enum ty name
+	ty: Ident,
+	// variant
+	variant: EnumVariantDispatch,
+}
+
+impl fmt::Debug for EnumVariantDispatchWithTy {
+	fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+		write!(f, "{}::{:?}", self.ty, self.variant)
+	}
+}
+
+impl ToTokens for EnumVariantDispatchWithTy {
+	fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
+		if let Some(inner) = &self.variant.inner {
+			let enum_name = &self.ty;
+			let variant_name = &self.variant.name;
+
+			let ts = quote! {
+				#enum_name::#variant_name(#inner::from(event))
+			};
+
+			tokens.extend(std::iter::once(ts));
+		}
+	}
+}
+
+/// An enum variant without the base type, contains the relevant inner type.
+#[derive(Clone)]
+struct EnumVariantDispatch {
+	/// variant name
+	name: Ident,
+	/// The inner type for which a `From::from` impl is anticipated from the input type.
+	/// No code will be generated for this enum variant if `inner` is `None`.
+	inner: Option<Type>,
+}
+
+impl fmt::Debug for EnumVariantDispatch {
+	fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+		write!(f, "{}(..)", self.name)
+	}
+}
+
+fn prepare_enum_variant(variant: &mut Variant) -> Result<EnumVariantDispatch> {
+	let skip = variant.attrs.iter().find(|attr| attr.path.is_ident("skip")).is_some();
+	variant.attrs = variant.attrs.iter().filter(|attr| !attr.path.is_ident("skip")).cloned().collect::<Vec<_>>();
+
+	let variant = variant.clone();
+	let span = variant.ident.span();
+	let inner = match variant.fields.clone() {
+		// look for one called inner
+		Fields::Named(FieldsNamed { brace_token: _, named }) if !skip => named
+			.iter()
+			.filter(
+				|field| {
+					if let Some(ident) = &field.ident {
+						ident.to_string() == "inner".to_owned()
+					} else {
+						false
+					}
+				},
+			)
+			.map(|field| Some(field.ty.clone()))
+			.next()
+			.ok_or_else(|| {
+				Error::new(span, "To dispatch with struct enum variant, one element must named `inner`")
+			})?,
+
+		// technically, if it has no inner types we cound not require the #[skip] annotation, but better make it consistent
+		Fields::Unnamed(FieldsUnnamed { paren_token: _, unnamed }) if !skip => unnamed
+			.first()
+			.map(|field| Some(field.ty.clone()))
+			.ok_or_else(|| Error::new(span, "Must be annotated with skip, even if no inner types exist."))?,
+		_ if skip => None,
+		Fields::Unit => {
+			return Err(Error::new(
+				span,
+				"Must be annotated with #[skip].",
+			))
+		}
+		Fields::Unnamed(_) => {
+			return Err(Error::new(
+				span,
+				"Must be annotated with #[skip] or have in `inner` element which impls `From<_>`.",
+			))
+		}
+		Fields::Named(_) => {
+			return Err(Error::new(
+				span,
+				"Must be annotated with #[skip] or the first wrapped type must impl `From<_>`.",
+			))
+		}
+	};
+
+	Ok(EnumVariantDispatch { name: variant.ident, inner })
+}
+
+fn impl_subsystem_dispatch_gen2(attr: TokenStream, item: TokenStream) -> Result<proc_macro2::TokenStream> {
```

Comment:
Please drop the 2 ;)

Reply:
Soo, the `2` was added to distinguish between `proc_macro2` and `proc_macro` API types, renamed to `_inner`.